core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -1408,4 +1404,239 @@ class ReplicaManagerTest {
+      val partition0Replicas01 = Seq[Integer](0, 1).asJava
+      val partition0Replicas10 = Seq[Integer](1, 0).asJava
+      val partition0Replicas1 = Seq[Integer](1).asJava
+
+      // make broker 0 the leader of partition 0
+      val leaderAndIsrRequest1 = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion,
+        controllerId, 0, brokerEpoch,
+        Seq(new LeaderAndIsrPartitionState()
+          .setTopicName(tp0.topic())
+          .setPartitionIndex(tp0.partition())
+          .setControllerEpoch(controllerEpoch)
+          .setLeader(0)
+          .setLeaderEpoch(leaderEpoch)
+          .setIsr(partition0Replicas0)
+          .setZkVersion(0)
+          .setReplicas(partition0Replicas0)
+          .setIsNew(true)).asJava,
+        Set(new Node(0, "host0", 0), new Node(1, "host1", 1)).asJava).build()
+
+      rm0.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest1, (_, _) => ())
+      rm1.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest1, (_, _) => ())
+
+      // Append a couple of messages.
+      for (i <- 1 to 2) {
+        appendRecords(rm0, tp0, TestUtils.singletonRecords(s"message $i".getBytes)).onFire { response =>
+          assertEquals(Errors.NONE, response.error)
+        }
+      }
+
+      // put rm0 and rm1 into reassignment state
+      leaderEpoch += leaderEpochIncrement
+      val leaderAndIsrRequest2 = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion, controllerId,
+        controllerEpoch, brokerEpoch,
+        Seq(new LeaderAndIsrPartitionState()
+          .setTopicName(tp0.topic())
+          .setPartitionIndex(tp0.partition())
+          .setControllerEpoch(controllerEpoch)
+          .setLeader(0)
+          .setLeaderEpoch(leaderEpoch)
+          .setIsr(partition0Replicas0)
+          .setZkVersion(0)
+          .setReplicas(partition0Replicas01)
+          .setAddingReplicas(partition0Replicas1)
+          .setIsNew(true)).asJava,
+        Set(new Node(0, "host0", 0), new Node(1, "host1", 1)).asJava).build()
+
+      rm0.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest2, (_, _) => ())
+      rm1.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest2, (_, _) => ())
+
+      assertEquals(1, rm0.reassigningPartitionsCount())
+      assertEquals(0, rm1.reassigningPartitionsCount())
+
+      // there should be no lag at this point as no data is fetched
+      assertEquals(0L, rm0.fetchersMaxLag())
+      assertEquals(0L, rm1.fetchersMaxLag())
+
+      // there should be no under-replicated partitions
+      assertEquals(0, rm0.underReplicatedPartitionCount)
+      assertEquals(0, rm1.underReplicatedPartitionCount)
+
+      // after the fetch the lag should be increased to 2 as 2 messages have been produced
+      val fetchResult = fetchAsFollower(rm0, tp0, new PartitionData(0, 0, Int.MaxValue, Optional.empty()))
+      var timeout = 1000L
+      while (!fetchResult.isFired && timeout > 0) {
+        wait(100L)
+        timeout -= 100
+      }
```

Comment:
idk about other people's preferences, but I always like to have such functionality inside `TestUtils`. Perhaps we can reuse `TestUtils.waitUntilTrue` and make it conditionally fail the test if the result isn't true (I assume we don't want to fail here)